### ISSUE TITLE
Eliminate warning when cloning a tensor using `torch.tensor(x)`

### ISSUE THREAD

**Issue by tshead2** [pytorch/pytorch#42188]
## 🚀 Feature

The following code:

```
import torch
a = torch.arange(0, 4)
b = torch.tensor(a)
```

generates a warning at the console:

```UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).```

This warning should be eliminated.

## Motivation

I often write code that is designed to accept any type that's convertible to a tensor as input, and I don't want my code to have any side effects on the input.  `torch.tensor()` provides the perfect semantics for this use-case since - as documented - it always copies its input.  For example:

```
def foo(x):
  x = torch.tensor(x)
  # Return something based on x
```

In this case, the warning is extremely annoying, since its suggestion to use `Tensor.clone().detach()` instead only works if the input is already a tensor.  Furthermore, it isn't clear why a warning is necessary here at all - it might as well be saying "Warning: this function works as documented!"

## Pitch

Please eliminate this warning.  If necessary, incorporate its advice into the documentation for `torch.tensor()` instead.

## Alternatives

Perhaps this warning is suggesting that `torch.tensor` will be deprecated or modified in the future?  If so, it would be better to incorporate those plans into the documentation.

**Comment by gchanan** [comment]
fwiw the documentation already covers the question I think:

> When data is a tensor x, torch.tensor() reads out ‘the data’ from whatever it is passed, and constructs a leaf variable. Therefore torch.tensor(x) is equivalent to x.clone().detach() and torch.tensor(x, requires_grad=True) is equivalent to x.clone().detach().requires_grad_(True). The equivalents using clone() and detach() are recommended.

**Comment by gchanan** [comment]
This part of the docs is pretty unclear:

> torch.tensor() always copies data. If you have a Tensor data and want to avoid a copy, use torch.Tensor.requires_grad_() or torch.Tensor.detach(). If you have a NumPy ndarray and want to avoid a copy, use torch.as_tensor().

Should say something like:
"If you want to avoid a copy, but set requires_grad, call..."
"If you want to wipe the autograd history..."

**Comment by gchanan** [comment]
I think the motivation for this was the unclear semantics around what happens with autograd.  I think we can at least not warn if:
- the tensor data does not have requires_grad set OR
- `requires_grad=False` is passed to `torch.tensor`

That would seem to give you a usable workaround for your case, as either the first condition is already true or you can pass the second condition manually.

**Comment by tshead2** [comment]
Agreed on most points, and I'll take what I can get.  I would still argue that a warning is a poor solution for what is either a documentation problem (you don't think the caller understands what the API will do) or a design problem (you don't think the API will do what the user actually wants).  Perhaps copy semantics need a rethink - as always, more explicit would be better.

Cheers,
Tim

**Comment by gchanan** [comment]
Both fair points.  It's definitely at least a documentation problem (as I noted above), but it's also true that people often don't read documentation unless you alert them that there is a potential issue (through error or warning).

The design problem is probably true too -- it's trying to provide a NumPy-like API in a case that NumPy doesn't have to worry about (what happens with autograd?).

**Comment by ezyang** [comment]
I support the `requires_grad=False` API idea, it seems pretty nice and explicit
